File: domains/games/apis/one_d4/src/main/java/com/muchq/games/one_d4/motifs/OverloadedPieceDetector.java
package com.muchq.games.one_d4.motifs;

import com.muchq.games.one_d4.engine.model.GameFeatures;
import com.muchq.games.one_d4.engine.model.Motif;
import com.muchq.games.one_d4.engine.model.PositionContext;
import java.util.ArrayList;
import java.util.List;

/**
 * Detects OVERLOADED_PIECE: a single defending piece that defends two or more friendly pieces that
 * are each under attack by the opponent. If that defender is used to defend one target, the other
 * target falls.
 *
 * <p>Detection (per position): for each piece of the side that just moved, find the opposing
 * (defending) side's pieces that are under attack. Then check if any single defending piece is the
 * sole or shared defender of two or more such attacked squares.
 */
public class OverloadedPieceDetector implements MotifDetector {

  @Override
  public Motif motif() {
    return Motif.OVERLOADED_PIECE;
  }

  @Override
  public List<GameFeatures.MotifOccurrence> detect(List<PositionContext> positions) {
    List<GameFeatures.MotifOccurrence> occurrences = new ArrayList<>();

    for (PositionContext ctx : positions) {
      if (ctx.lastMove() == null) continue;

      String placement = ctx.fen().split(" ")[0];
      int[][] board = PinDetector.parsePlacement(placement);

      // After the move, the attacker is !ctx.whiteToMove() and the defender is ctx.whiteToMove()
      boolean attackerIsWhite = !ctx.whiteToMove();
      boolean defenderIsWhite = ctx.whiteToMove();

      int[] overloaded = findOverloadedPiece(board, attackerIsWhite, defenderIsWhite);
      if (overloaded != null) {
        int r = overloaded[0], c = overloaded[1];
        String target = BoardUtils.pieceNotation(board[r][c], r, c);
        GameFeatures.MotifOccurrence occ =
            GameFeatures.MotifOccurrence.withPiece(
                ctx, "Overloaded piece at move " + ctx.moveNumber(), null, target);
        if (occ != null) occurrences.add(occ);
      }
    }

    return occurrences;
  }

  /**
   * Returns {r, c} of the first overloaded defending piece found, or null if none. An overloaded
   * piece defends 2+ squares that are each under attack.
   */
  private int[] findOverloadedPiece(
      int[][] board, boolean attackerIsWhite, boolean defenderIsWhite) {
    // Collect squares with defending-side pieces that are under attack by the attacking side
    List<int[]> attackedDefendingSquares = new ArrayList<>();
    for (int r = 0; r < 8; r++) {
      for (int c = 0; c < 8; c++) {
        int piece = board[r][c];
        if (piece == 0) continue;
        if ((piece > 0) != defenderIsWhite) continue;
        if (BoardUtils.countAttackers(board, r, c, attackerIsWhite) > 0) {
          attackedDefendingSquares.add(new int[] {r, c});
        }
      }
    }

    if (attackedDefendingSquares.size() < 2) return null;

    // For each defending piece, count how many attacked defending squares it can recapture on
    for (int r = 0; r < 8; r++) {
      for (int c = 0; c < 8; c++) {
        int piece = board[r][c];
        if (piece == 0) continue;
        if ((piece > 0) != defenderIsWhite) continue;

        int defended = 0;
        for (int[] sq : attackedDefendingSquares) {
          if (sq[0] == r && sq[1] == c) continue; // piece can't defend itself this way
          if (BoardUtils.pieceAttacks(board, r, c, sq[0], sq[1])) {
            defended++;
          }
        }
        if (defended >= 2) return new int[] {r, c};
      }
    }
    return null;
  }
}

File: domains/games/libs/one_d4_motifs/overloaded_piece_detector.cc
// OVERLOADED_PIECE: a defender doing two jobs it cannot both keep.
//
// Declared, queryable, and always empty until now — BoardUtils has no
// notion of a defender at all, so the Java pipeline had nothing to build it
// from. attackers(board, color, square) asked of both colors gives it
// directly: for any square, who attacks it and who defends it.
//
// A piece is overloaded when two or more of our pieces are attacked and it
// is the only thing defending each. Take one and the other falls.
// Deliberately not "defends two things", which is most pieces in most
// positions and would say nothing.

#include <memory>
#include <optional>
#include <utility>
#include <vector>

#include "absl/strings/str_cat.h"
#include "chess.hpp"
#include "domains/games/libs/chess_cpp/board_facts.h"
#include "domains/games/libs/chess_cpp/replay.h"
#include "domains/games/libs/chess_cpp/side.h"
#include "domains/games/libs/one_d4_motifs/board_scan.h"
#include "domains/games/libs/one_d4_motifs/detector.h"
#include "domains/games/libs/one_d4_motifs/detectors.h"
#include "domains/games/libs/one_d4_motifs/notation.h"

namespace one_d4 {
namespace {

using chess_cpp::Position;
using chess_cpp::Side;

/// A piece, the attacked squares it alone defends, and everything of
/// theirs bearing on those squares.
struct Duties {
  chess::Square defender;
  std::vector<chess::Square> defended;
  chess::Bitboard attackers;
};

class OverloadedPieceDetector : public Detector {
 public:
  Motif motif() const override { return Motif::kOverloadedPiece; }

  // Attributed to the side that just moved — the one with something to
  // win — while the overloaded piece itself belongs to the side to move.
  void OnPosition(const Position& position, Findings& out) override {
    const chess::Board& board = position.board;
    const Side defending = position.side_to_move;
    const Side attacking = chess_cpp::Opponent(defending);

    // For each of our pieces, the attacked squares it alone defends.
    std::vector<Duties> duties;
    ForEachSquare([&](int row, int col) {
      const chess::Square square = SquareAt(row, col);
      if (!BelongsTo(board.at(square), defending)) return;
      // A king in check is not a piece being held; without this every check
      // with a single defender on the board reads as an overload, and a
      // knight fork reads as one twice.
      if (board.at(square).type() == chess::PieceType::KING) return;
      // The defence has to be holding something back. A pawn attacked by a
      // queen is not defended in any meaningful sense — taking it loses the
      // queen — and counting those makes half the pieces on the board
      // overloaded.
      const std::optional<int> cheapest = CheapestAttackerValue(board, attacking, square);
      if (!cheapest.has_value() || *cheapest > Value(board.at(square))) return;

      const chess::Bitboard defenders = chess_cpp::facts::AttackersOf(board, defending, square);
      if (defenders.count() != 1) return;

      const chess::Square defender(defenders.lsb());
      const chess::Bitboard attackers = chess_cpp::facts::AttackersOf(board, attacking, square);
      for (Duties& duty : duties) {
        if (duty.defender == defender) {
          duty.defended.push_back(square);
          duty.attackers |= attackers;
          return;
        }
      }
      duties.push_back({defender, {square}, attackers});
    });

    for (const auto& [defender, defended, attackers] : duties) {
      if (defended.size() < 2) continue;
      // Two jobs and one attacker is not an overload: it can cash one
      // square, the defender recaptures, and the other square is no longer
      // attacked. Pawn tension reads as an overload without this.
      if (attackers.count() < 2) continue;
      // One row per piece it cannot keep, so the read path can name both
      // halves of the tactic.
      for (const chess::Square square : defended) {
        Finding finding;
        finding.description = absl::StrCat("Overloaded piece at move ", position.move_number);
        finding.attacker = PieceNotation(board.at(defender), defender);
        finding.target = PieceNotation(board.at(square), square);
        out.Add(position, std::move(finding));
      }
    }
  }
};

}  // namespace

std::unique_ptr<Detector> MakeOverloadedPieceDetector() {
  return std::make_unique<OverloadedPieceDetector>();
}

}  // namespace one_d4
